### PR TITLE
Refactor fold_arguments and add typing

### PR DESCRIPTION
**Pull request by sklam**
Separate out different uses of fold_arguments into `SymbolFolder`, `TypeFolder` and `LowerFolder`

### CODE REVIEW

**Comment by sklam** [comment]
Trying a typeguard run in `typeguard_mypy_50`

**Comment by sklam** [comment]
This PR will help debugging and fixing: https://github.com/numba/numba/issues/7933

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch and for putting in the large amount of effort required to regularise this pattern in the code base, I think this is a considerable improvement. I've given this a review and have left comments inline, most of which are around adding docstrings/commentary and a couple about whether whether the handlers need to be more than just a singleton? Once addressed should be good to merge. Thanks again!

File: numba/core/dispatcher.py [hunk 1]
```diff
@@ -98,27 +134,18 @@ def __init__(self, py_func, targetdescr, targetoptions, locals,
         # the exceptions.
         self._failed_cache = {}
 
-    def fold_argument_types(self, args, kws):
+    def fold_argument_symbol(self, args: pt.Tuple[ir.Var, ...], kws):
+        args = FoldArguments(self.pysig, SymbolFolder()).fold(args, kws)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/dispatcher.py:138]: It's acting like a "vtable". I'm not sure what it'd gain to make it a singleton-like. Wouldn't a singleton allow for leaking state?

File: numba/core/dispatcher.py [hunk 2]
```diff
@@ -98,27 +134,18 @@ def __init__(self, py_func, targetdescr, targetoptions, locals,
...
-
-        def stararg_handler(index, param, values):
-            return types.StarArgTuple(values)
-        # For now, we take argument values from the @jit function, even
-        # in the case of generated jit.
-        args = fold_arguments(self.pysig, args, kws,
-                              normal_handler,
-                              default_handler,
-                              stararg_handler)
+        args = SymbolFoldArguments(self.pysig, TypeFolder()).fold(args, kws)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/dispatcher.py:148]: oh... hm.... good catch!

File: numba/core/dispatcher.py [hunk 3]
```diff
@@ -84,6 +86,40 @@ def _numba_type_(self):
         return types.Omitted(self.value)
 
 
+class TypeFolder:
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/dispatcher.py:89]: Yes on ABC. Not sure why to make it a `namedtuple`

**Comment by sklam** [comment]
Note to self: 

- related issue and patch https://github.com/numba/numba/issues/8047#issuecomment-1125526561

**Pull request closed without merging**